### ISSUE TITLE
Velero install error azure.BearerAuthorizer Failed to refresh the Token for request to https://management.azure.com

### ISSUE THREAD

**Issue by ezYakaEagle442** [vmware-tanzu/velero-plugin-for-microsoft-azure#46]
I run AKS 1.16.7 with --enable-managed-identity.
After Velero 1.3.2 Install with CLI I hit this issue :

An error occurred: some backup storage locations are invalid: error getting backup store for location "default": rpc error: code = Unknown desc = azure.BearerAuthorizer#WithAuthorization: **Failed to refresh the Token for request to** https://management.azure.com/subscriptions/XXXXX/resourceGroups/rg-backup-strcsi-francecentral/providers/Microsoft.Storage/storageAccounts/velero45edfb8d58f0/listKeys?api-version=2018-02-01: **StatusCode=401** -- Original Error: adal: Refresh request failed. Status Code = '401'. Response body: {"error":"invalid_client","error_description":"AADSTS7000215: Invalid client secret is provided.\r\nTrace ID: 4b050253-cb2f-4ed5-b841-xxxxx\r\nCorrelation ID: e094ec06-4a6b-4750-9b6c-e1c56fa63457\r\nTimestamp: 2020-05-24 18:47:35Z","error_codes":[7000215],"timestamp":"2020-05-24 18:47:35Z","trace_id":"4b050253-cb2f-4ed5-b841-6c200e9c6900","correlation_id":"e094ec06-4a6b-4750-9b6c-e1c56fa63457","error_uri":"https://login.microsoftonline.com/error?code=7000215"}

It looks like a very similar issue to **https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/issues/17** , How to check that Velero can read correctly the credential file ./credentials-velero ?

I have performed the pre-requisites below :

```sh

# Create the storage account.

AZURE_STORAGE_ACCOUNT_ID="velero$(uuidgen | cut -d '-' -f5 | tr '[A-Z]' '[a-z]')"
echo "AZURE_STORAGE_ACCOUNT_ID" $AZURE_STORAGE_ACCOUNT_ID

az storage account create \
    --name $AZURE_STORAGE_ACCOUNT_ID \
    --kind BlobStorage \
    --sku Standard_GRS \
    --encryption-services blob \
    --https-only true \
    --resource-group $rg_name \
    --access-tier Hot

BLOB_CONTAINER=velero
az storage container create -n $BLOB_CONTAINER --public-access off --account-name $AZURE_STORAGE_ACCOUNT_ID

velero_sp_password=$(az ad sp create-for-rbac --name velero-$appName --role Contributor --query password -o tsv)

echo "Velero Service Principal PWD" $velero_sp_password

#velero_sp_id=$(az ad sp list --all --query "[?appDisplayName=='velero'].{appId:appId}" --output tsv)
#velero_sp_id=$(az ad sp list --show-mine --query "[?appDisplayName=='velero'].{appId:appId}" --output tsv)
velero_sp_id=$(az ad sp list --display-name velero-$appName --query '[0].appId' -o tsv)
echo "Velero Service Principal ID:" $velero_sp_id 
az ad sp show --id $velero_sp_id

tenantId=$(az account show --query tenantId -o tsv)

aks_client_id=$(az aks show -g $rg_name -n $cluster_name --query identityProfile.kubeletidentity.clientId -o tsv)
echo "AKS Cluster Identity Client ID " $aks_client_id

cat << EOF > ./credentials-velero
AZURE_SUBSCRIPTION_ID=${subId}
AZURE_TENANT_ID=${tenantId}
AZURE_CLIENT_ID=${aks_client_id}
AZURE_CLIENT_SECRET=${velero_sp_password}
AZURE_RESOURCE_GROUP=${rg_name}
AZURE_CLOUD_NAME=AzurePublicCloud
EOF

cat ./credentials-velero

velero install \
    --provider azure \
    --plugins velero/velero-plugin-for-microsoft-azure:v1.0.1 \
    --bucket $BLOB_CONTAINER \
    --secret-file ./credentials-velero \
    --backup-location-config resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,storageAccount=$AZURE_STORAGE_ACCOUNT_ID,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID \
    --snapshot-location-config apiTimeout=$SNAPSHOT_TIMEOUT,resourceGroup=$AZURE_BACKUP_RESOURCE_GROUP,subscriptionId=$AZURE_BACKUP_SUBSCRIPTION_ID

for pod in $(k get po -l "app.kubernetes.io/name=velero" -n $velero_ns -o=name)
do
  k describe $pod -n  $velero_ns # | grep -i "Error"
  k logs $pod $velero_ns #| grep -i "Error"
done

```

**Comment by ezYakaEagle442** [comment]
Note: during the SP creation, the --scope was not explicitly set however by default it does apply to subscription root.

az ad sp create-for-rbac --help

    --scopes           : Space-separated list of scopes the service principal's role assignment
                         applies to. Defaults to the root of the current subscription

**Comment by ezYakaEagle442** [comment]
I just see today the issue, there is mistmatch in the Client Secret & Password ! : 

AZURE_CLIENT_ID=${aks_client_id}
AZURE_CLIENT_SECRET=**${velero_sp_password}**

I can see correctly the secret :
```sh
for secret in $(k get secrets -n $velero_ns -o custom-columns=:metadata.name)
do
  #k describe secret $secret -n $velero_ns
  echo -e "\n"
  # velero_secret=$(k get secret $secret -n $velero_ns -o jsonpath="{.data.token}" | base64 --decode)
  velero_secret=$(k get secret $secret -n $velero_ns -o jsonpath="{.data.cloud}" | base64 --decode)
  echo "Velero secret " 
  echo -e "\n"
  echo $velero_secret
done
```

so finally the question is how to configure Velero when Managed Identies is enables in AKS ?

See :

- [https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/issues/11](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/issues/11)
- [https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/issues/12](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure/issues/12)

-

**Comment by skriss** [comment]
@ezYakaEagle442 I'm not entirely clear on if this issue should be resolved now that #51 is merged - if it is, feel free to close this out, and if not, let us know :)

**Comment by ashish-amarnath** [comment]
@ezYakaEagle442 I am going to close this issue for inactivity. But, please reopen or reach out if you have more questions.
